### ISSUE TITLE
loss=nan in training after adding "--max-source/target-positions 2k" to my working model

### ISSUE THREAD

**Issue by gorka96** [pytorch/fairseq#2260]
## ❓ Questions and Help

### Before asking:   
1. search the issues. DONE
2. search the docs. DONE

#### What is your question?

I have a working transformer model, for a document level seq2seq simple task. I built a baseline to begin, with generic hiperparameters taken from a NMT example, which works quite well (I need to add BPE and more data, and tune training hyperparameters to improve it, but the model trains and generated somehow the expected output).

But as it is a document to document task, I need to increase the max seq length from 1024 to at least 2-3K (max length of the valid/test dataset), and ideally to 4k.

I understand, from the documentation and some answers here, that that this is achieved using --max-source-positions and --max-target-positions in the default --task translation.

But when I try to train this model, my loss becomes in "nan" from the 1st batch.

I don't know if i'm doing something wrong, I didn't understood how to use --max-target-positions correctly, I have to tune the hyperparameters of the transformer after increasing the length, or what's causing this nan loss problem.

Thaks for taking the time to help on this.

Gorka

#### Code

THIS IS WORKING CORRECTLY:

```
fairseq-preprocess --source-lang src --target-lang trg --trainpref data/train_ARRAU_seq --validpref data/dev_ARRAU
 --testpref data/test_ARRAU --destdir data-bin/text2cor.seq --workers 8

fairseq-train data-bin/text2cor.seq --arch transformer_iwslt_de_en --optimizer adam --adam-betas '(0.9, 0.98)' 
--clip-norm 0.1 --lr 5e-4 --lr-scheduler inverse_sqrt --warmup-updates 4000 --dropout 0.3 
--weight-decay 0.0001 --criterion label_smoothed_cross_entropy --label-smoothing 0.1 --no-epoch-checkpoints 
--num-workers 8 --max-epoch 10 --save-dir checkpoints/transformer_text2cor_seq 
--skip-invalid-size-inputs-valid-test --max-tokens 4096 --update-freq 8

fairseq-generate data-bin/text2cor.seq --skip-invalid-size-inputs-valid-test 
--path checkpoints/transformer_text2cor_seq/checkpoint_best.pt --beam 5
```

THIS IS WHERE I GOT THE NAN LOSS PROBLEM:

`fairseq-train data-bin/text2cor.seq --arch transformer_iwslt_de_en --optimizer adam --adam-betas '(0.9, 0.98)' 
--clip-norm 0.1 --lr 5e-4 --lr-scheduler inverse_sqrt --warmup-updates 4000 --dropout 0.3 
--weight-decay 0.0001 --criterion label_smoothed_cross_entropy --label-smoothing 0.1 --no-epoch-checkpoints 
--num-workers 8 --max-epoch 10 --save-dir checkpoints/transformer_text2cor_seq 
--skip-invalid-size-inputs-valid-test --max-tokens 4096 --update-freq 8 
--max-source-positions 2048 --max-target-positions 2048`

...
| [src] dictionary: 18264 types
[trg] dictionary: 1728 types
| loaded 18 examples from: data-bin/text2cor.seq/valid.src-trg.src
| loaded 18 examples from: data-bin/text2cor.seq/valid.src-trg.trg
| data-bin/text2cor.seq valid src-trg 18 examples`
...
| model transformer_iwslt_de_en, criterion LabelSmoothedCrossEntropyCriterion                                                                                           | num. model params: 42663936 (num. trained: 42663936)                                                                                                                  | training on 1 GPUs                                                                                                                                                    | max tokens per GPU = 4096 and max sentences per GPU = None                                                                                                            | no existing checkpoint found checkpoints/transformer_text2cor_seq2k/checkpoint_last.pt                                                                                | loading train data for epoch 0                                                                                                                                        | loaded 609781 examples from: data-bin/text2cor.seq/train.src-trg.src                                                                                                  | loaded 609781 examples from: data-bin/text2cor.seq/train.src-trg.trg                                                                                                  | data-bin/text2cor.seq train src-trg 609781 examples
| WARNING: 11644 samples have invalid sizes and will be skipped, max_positions=(2048, 2048), first few sample ids=[96983, 96942, 96984, 97067, 97027, 96943, 96944, 96902, 96858, 96945]                                                                                                                                                        | NOTICE: your device may support faster training with --fp16                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
| epoch 001:   0%| | 1/12560 [00:01<5:52:58,  1.69s/it, **loss=nan, nll_loss=nan, ppl=nan**, wps=3122, ups=0, wpb=27607.000, bsz=35.000, num_updates=1, lr=1.25e-07, gnorm=n                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                
| epoch 001:   0%| | 2/12560 [00:02<5:16:01,  1.51s/it, **loss=nan, nll_loss=nan**, ppl=nan, wps=24736, ups=1, wpb=27385.500, bsz=38.500, num_updates=2, lr=2.5e-07, gnorm=n                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        
| epoch 001:   0%| | 3/12560 [00:03<4:49:35,  1.38s/it, l**oss=nan, nll_loss=nan, ppl=nan**, wps=25190, ups=1, wpb=27569.000, bsz=39.333, num_updates=3, lr=3.75e-07, gnorm=n

#### What have you tried?

I tried training with an smaller --lr, different --max-tokens and --update-freq, and --fp16 settings, but I get the same issue.

#### What's your environment?

 - fairseq Version: 0.9.0
 - PyTorch Version: 1.6
 - OS (e.g., Linux): Linux
 - How you installed fairseq (`pip`, source): I would have to ask it
 - Build command you used (if compiling from source): I would have to ask it
 - Python version: 3.6.8
 - CUDA/cuDNN version: I would have to ask it
 - GPU models and configuration: 1 x Rtx2080Ti

**Comment by alexeib** [comment]
can you look at the actual input and target samples you are using and see if they make sense? if everything works with fewer max-positions and stops working when you increase it, its very likely a data problem

**Comment by gorka96** [comment]
i tried to train with 1025 max-positions, and i already got the loss=nan problem, so next, I will check the data as you suggested, to see if that the source of the problem.

Thanks.

**Comment by gorka96** [comment]
Hi,

I filtered the dataset into sequences shorter than 1024 (1020 to avoid problems), and it works properly.

Then, I added a single instance of length 1025 on target (source sequences are a bit shorter at this task (coreference)). And when I trained the system with **--max-source-positions 2048--max-target-positions 2048** , loss converted into nan when the long sequence is taken for training. 

So now we know that that single training instance is causing the problem. I open it, and there is nothing strange about it. it looks well formed, the vocab used in source is shared with the rest of training data as I am augmenting the training set with diferent sequences of diferent lengths from the same documents, so the tokens are repeated. about the target sequence, its a representation of the coreference anotation, created by my code, which should'nt give any problem.

Here it's the source sequence:

`but I do n't he 's wearing like an apron with huge pockets . But I do n't think you see the apron at first . I do n't know if that 's important or not . And uh and then he gets down out of the tree . and he dumps ail pears into the basket and the basket 's full . and one of the pears drops down to the floor . and he picks it up . and he takes kerchief off . and he wipes it off . and places it in in the basket which is very full . That 's why it fell off in the first place . And um then he climbs back up the ladder . and he and he starts picking pears again . And then while he 's up in the ladder . let 's see is it while he 's up in the ladder . Or or before . Um anyway . a guy comes by leading a goat . And the goat 's aaarrr but and they do n't talk to each other . they do n't I do n't think they even look at each other . and the guy walks by . And you watch the goat . disappearing all the way . and then then you 're back to the man in the tree . And he 's up there in the tree . And then definitely when he 's up there . a kid comes by on a bicycle . From the direction where the goat man left . okay . And Um the bicycle 's way too big for the kid . I 'm giving you all these details . I do n't know if you want them . Um the the reason I 'm giving you the details is cause I do n't know what the point of the movie was . Okay . So maybe you can see something that I did n't . Okay . Um g a kid comes by on a bicycle . he stops . he gets off bike . um the movie was in color . And the movie had sound track . It 's important . And then the mo the whole movie started with a a cock crowing . And then you see . Anyway . I just remembered that . Anyway . so um the kid on the bicycle . gets off the bicycle . I think he he picks up the b-asket of pears and he puts it on blcycle . on the front . And he rides off . And you see him . riding off and the next scene you see him . like at a at a distance . a pan shot . And he 's riding on a lane . riding on a bicycle and they cro you know . Cross paths . And he kind of watches her go by . And because he 's watching her . he when he turns around hat flies off . Okay . And then he bumps into a rock . And u m falls down . and the pears all spray all over the place . and and then you see him fall down . and he gets up . and brushes himself off . and he goes you know and then you see three other boys about age . He 's like maybe what . ten or something . Twelve . I do n't know . And and you look at them . and and they see him . and they come up . and without saying anything . there 's no speech in the whole movie. . without saying anything . they um help him put the pears back in the basket . Only they do n't wipe them off first . Just pick them up . and pour them in . And theyhelp him get bike up . and one kid takes the rock that he tripped on and he throws it off to the side of the road . and and they set him upright . and they and then he you know they I don1they do n't say thank you . he just splits . And then they walk the three boys walk down the road and they see the kid 's hat . And so one of them whistles . And the kid on the bike turns around . and then he brings hat to him . and the g guy on the bike gives the other kids gives the kid that returns hat three pears to share with buddies . So and then he splits . With hat . with pears . everything . And um I just remembered another detail . Before he meets the girl . One of the pears . 1t 's a bumpy road little dirt road . one of the pears flies out . Whatever . And Um then you do n't see the kid with the bike with the pears anymore . And the three kids that are walking . one of them has a what do you call those little um paddleball . You know . with a ball on a uh chong chong chong chong . So they 're walking along . and they brush off pears . and they start eating it . Then they walk by the man who was picking the pears . Who looks like a Mexican-American if that 's important . Um and he 's ge he 's getting down out of the tree and fucking with the ladder . And he sees and and there 's one full basket of pears there . and and an empty basket . And the other full basket is gone .`

here the target sequence:

`_ (0) _ _ (1) _ _ _ (2 _ _ (3 2) | 3) _ _ (0) _ _ _ (4) _ (2 2) _ _ _ (0) _ _ _ _ (5) _ _ _ _ _ _ _ _ _ (1) _ _ _ _ (6 6) _ _ (1) _ (7 7) | (8) _ (9 9) _ (9 9) _ _ _ _ (10 _ (8 10) | 8) _ _ _ (11 11) _ _ (1) _ (10) _ _ _ (1) _ (12) _ _ _ (1) _ (10) _ _ _ _ (10) _ _ (9 _ _ _ _ 9) _ (13) _ _ (10) _ _ _ (14 _ 14) _ _ _ _ (1) _ _ _ (15 15) _ _ (1) _ (1) _ _ (16) _ _ _ _ _ (1) _ _ _ (15 15) _ _ _ _ _ (17) _ (1) _ _ _ (15 15) _ _ _ _ _ _ _ _ (18 18) _ _ _ (19 19) _ _ (19 19) _ _ _ _ (20) _ _ _ _ (21 21) _ (20) _ _ (0) _ _ _ (20) _ _ _ (22 22) _ _ (18 18) _ _ _ _ (23) _ (19 19) _ _ (24 _ 24) _ _ _ _ (25) _ _ _ (1 _ _ (6 1) | 6) _ _ (1) _ _ (26 _ (6 26) | 6) _ _ _ _ _ (1) _ _ (26) _ (27 27) _ _ _ (28 28) _ _ (29 _ _ (18 (19) 18) 29) _ _ _ _ _ (28 28) _ _ _ _ _ (27 27) _ (0) _ _ (30) (31 (32 31) | 32) _ (0) _ _ _ _ (30) _ (32) _ _ _ (33 _ (0) _ _ (30) (32 33) | 32) _ _ (0) _ _ _ _ (34 _ _ (35 34) | 35) _ _ _ _ _ _ (30) _ _ (36 _ (0) _ 36) _ _ _ _ _ (27 27) _ _ _ (28 28) _ (27) _ _ (27) _ _ (28) _ _ (35 35) _ _ (37) _ _ (35 35) _ (38 38) _ (39) _ _ _ _ _ (35 _ _ _ 35) _ _ _ (40 40) _ _ _ _ (41) _ _ _ _ (0) _ _ (42) _ _ _ _ _ (27 _ _ (28 27) | 28) _ _ _ (28 28) _ (0) _ (27) (27) _ _ (9 _ _ 9) | (16) _ (27) _ (9) _ (28) _ _ (43 43) _ _ (27) _ _ _ _ (44) _ (27) _ _ _ _ (45 _ 45) (46) _ (27) _ _ _ _ _ (47 47) _ _ (48 48) _ _ (27) _ _ _ (49 49) _ _ _ (28 28) _ (50) _ (51) _ _ _ (52) _ _ (27) _ _ _ (53) _ _ _ _ _ (27) _ _ (53) _ (27) _ (27) _ _ (54) _ _ _ _ _ _ _ (27) _ _ (55 55) _ _ _ _ _ _ _ _ (9 9) _ _ _ _ (56 56) _ _ _ _ (57) _ (27) _ _ _ _ (27) _ _ _ _ _ (27) _ _ _ (27) _ (58) _ _ _ (59) _ (60 _ _ _ 60) | (61) _ (27) _ _ _ _ _ (62 | (63) _ 62) | (64) _ (65) _ (0) _ _ _ _ _ _ (66) _ _ (60) _ _ _ (60) _ (27) _ _ (60) _ _ _ _ _ _ (67) _ (68) _ (69 69) _ (35 _ 35) _ _ _ (70) _ (60) _ _ (27) _ (9 9) _ _ (9 9) _ _ (60) _ _ _ (9) _ _ _ _ _ (9) _ _ _ _ (9) _ _ _ _ (27) _ (28) _ _ _ (71 71) _ (55 _ _ (27) _ 55) _ (71) _ (55) _ _ (72 _ _ (49 72) | 49) _ _ _ (60) _ (27) _ _ _ (60) _ _ (27) (73) _ (60) (0) _ _ _ _ _ _ _ (27) _ _ _ _ _ (60) _ (60 _ 60) _ _ (49 49) _ (60) _ (54 | (27 _ 27) 54) _ _ _ (74 _ 74) | (60) _ _ _ (27 _ _ (28 27) | 28) _ _ _ _ _ (74) _ (54) _ (27) _ _ (27 _ _ _ (28 27) | 28) _ (60 _ 60) _ (74 _ _ _ 74) | (54) (75 75) _ _ _ (76) _ _ _ _ (27) _ _ _ (54) _ _ (9) _ (77) _ _ _ (0) _ _ (78 78) _ _ (27) _ (53 53) _ (79 _ (9 79) | 9) _ (49) _ (80 _ 80) (81 _ 81) _ (79 _ (9 79) | 9) _ _ _ (82) _ _ _ _ (83) _ _ _ (27 _ _ (28 28) _ (9 27) | 9) _ _ _ (60 _ _ _ _ 60) _ (84 _ 84) | (60) _ _ _ _ (85) _ (86 _ _ 86) _ (87) _ _ _ (88 _ _ (89 _ _ _ _ 88) | 89) _ _ (60) _ _ _ _ _ (60) _ _ (75) _ _ (60) _ _ (75) _ _ (60) _ _ (1 _ _ _ _ (9 1) | 9) _ (1) _ _ (90 90) _ (91) _ _ _ _ _ (1) _ _ (1) _ _ _ _ _ (6 6) _ _ _ (15 15) _ _ (1) _ _ _ (92) _ (93 | (94 _ _ _ 93) | (16) (95) _ _ _ (96 _ 94) | 96) _ _ (9 _ _ 9) _ _ _`

does anyone find something suspicious in the data? or may it be something else the cause of my issue?
